rfc/rfc-51/rfc-51.md

```diff
@@ -215,18 +245,31 @@ Note:
 
 - Only instants that are active can be queried in a CDC scenario.
 - `CDCReader` manages all the things on CDC, and all the spark entrances(DataFrame, SQL, Streaming) call the functions in `CDCReader`.
-- If `hoodie.table.cdc.supplemental.logging` is false, we need to do more work to get the change data. The following illustration explains the difference when this config is true or false.
+- If `hoodie.table.cdc.supplemental.logging.mode=KEY_OP`, we need to compute the changed data. The following illustrates the difference.
 
 ![](read_cdc_log_file.jpg)
 
 #### COW table
 
-Reading COW table in CDC query mode is equivalent to reading a simplified MOR table that has no normal log files.
+Reading COW tables in CDC query mode is equivalent to reading MOR tables in RO mode.
 
 #### MOR table
 
-According to the design of the writing part, only the cases where writing mor tables will write out the base file (which call the `HoodieMergeHandle` and it's subclasses) will write out the cdc files.
-In other words, cdc files will be written out only for the index and file size reasons.
+According to the section "Persisting CDC in MOR", CDC data is available upon base files' generation.
+
+When users want to get fresher real-time CDC results:
+
+- users are to set `hoodie.datasource.query.incremental.type=snapshot`
+- the implementation logic is to compute the results in-flight by reading log files and the corresponding base files (
+  current and previous file slices).
+- this is equivalent to running incremental-query on MOR RT tables
+
+When users want to optimize compute-cost and are tolerant with latency of CDC results,
+
+- users are to set `hoodie.datasource.query.incremental.type=read_optimized`
+- the implementation logic is to extract the results by reading persisted CDC data and the corresponding base files (
+  current and previous file slices).
```

Comment:
Do we need this two config options then ? 
`hoodie.datasource.query.incremental.type=snapshot`
`hoodie.datasource.query.incremental.type=read_optimized`

Very confusing from my side, shouldn't it always be reading the refresh cdc data here ? Why we expose a ro view in cdc streaming read, for what use case ?

Reply:
I agree. We can skip providing read optimized CDC - not easy to explain/understand. We can mention in the design that we can either compute the CDC on the fly or precompute it during the write/compaction. CDC on the fly is something we dont have to support in the implementation right away I think.

Reply:
There are indexing choices which result in eventual CDC stream freshness and we could put the impl of constructing CDC on the read to later point

Reply:
> it is robust fallback for the writer anomalies.

can you pls clarify what cases exactly represent writer anomalies? if write failed, data and CDC data should be discarded and commit should abort. In case of data corruption, computing with whatever data persisted there anyway won't be reliable, will it? i think freshness is instead the main reason for doing CDC on the fly.

As for the plan to support read mode, I discussed with @YannByron that computing CDC on the fly (`snapshot` mode) is an easier first implementation step. 2nd step to support `read_optimized` is meant for customers with no strong freshness requirement and want to save compute costs pulling CDC.

Reply:
> can you pls clarify what cases exactly represent writer anomalies

I'm a little worried about the cdc semantics correctness before it is widely used in production, such as the deletion, the merging, the data sequence.

> 2nd step to support read_optimized i

We can all it an optimization, but not exposed as read_optimized, which conflicts a little with current read_optimized view, WDYT.

Reply:
@danny0405 i agree on the design perspective, it's ideal that we don't need to let users configure this. Compaction generated CDC data should be stitched with on-the-fly inferred CDC data to optimize perf automatically. @YannByron did mention some implementation challenges. But from RFC standpoint, we should aim for the better design and strive for ways to achieve it.

Reply:
chatted with @YannByron and @danny0405 and we aligned on removing the config and allow cdc query to merge on-the-fly result with compacted results, as an optimized approach.